src/Fantomas/AstTransformer.fs

```diff
@@ -62,1635 +34,1170 @@ module private Ast =
+                [ mkNode SynModuleDecl_ModuleAbbrev range ]
+                |> finalContinuation
+            | SynModuleDecl.NestedModule (sci, _, decls, _, range) ->
+                let continuations : ((TriviaNodeAssigner list -> TriviaNodeAssigner list) -> TriviaNodeAssigner list) list =
+                    decls |> List.map visit
+
+                let finalContinuation (nodes: TriviaNodeAssigner list list) : TriviaNodeAssigner list =
+                    [ mkNode SynModuleDecl_NestedModule range
+                      yield! visitSynComponentInfo sci
+                      yield! (List.collect id nodes) ]
+                    |> finalContinuation
+
+                Continuation.sequence continuations finalContinuation
+            | SynModuleDecl.Let (_, bindings, range) ->
+                mkNode SynModuleDecl_Let range
+                :: (bindings |> List.collect visitSynBinding)
+                |> finalContinuation
+            | SynModuleDecl.DoExpr (_, expr, range) ->
+                mkNode SynModuleDecl_DoExpr range
+                :: visitSynExpr expr
+                |> finalContinuation
+            | SynModuleDecl.Types (typeDefs, range) ->
+                mkNode SynModuleDecl_Types range
+                :: (typeDefs |> List.collect visitSynTypeDefn)
+                |> finalContinuation
+            | SynModuleDecl.Exception (exceptionDef, range) ->
+                mkNode SynModuleDecl_Exception range
+                :: (visitSynExceptionDefn exceptionDef)
+                |> finalContinuation
+            | SynModuleDecl.Open (target, parentRange) ->
+                // we use the parent ranges here to match up with the trivia parsed
+                match target with
+                | SynOpenDeclTarget.ModuleOrNamespace (_, _range) ->
+                    mkNode SynModuleDecl_Open parentRange
+                    |> List.singleton
+                    |> finalContinuation
+                | SynOpenDeclTarget.Type (synType, _range) ->
+                    mkNode SynModuleDecl_OpenType parentRange
+                    :: (visitSynType synType)
+                    |> finalContinuation
+            | SynModuleDecl.Attributes (attrs, range) ->
+                mkNode SynModuleDecl_Attributes range
+                :: (visitSynAttributeLists range attrs)
+                |> finalContinuation
+            | SynModuleDecl.HashDirective (hash, range) ->
+                [ mkNode SynModuleDecl_HashDirective range
+                  visitParsedHashDirective hash ]
+                |> finalContinuation
+            | SynModuleDecl.NamespaceFragment (moduleOrNamespace) ->
+                visitSynModuleOrNamespace moduleOrNamespace
+                |> finalContinuation
+
+        visit ast id
+
+    and visitSynExpr (synExpr: SynExpr) : TriviaNodeAssigner list =
+        let rec visit
+            (synExpr: SynExpr)
+            (finalContinuation: TriviaNodeAssigner list -> TriviaNodeAssigner list)
```

Comment:
By the way, I find these things easier to write correctly if I stay generic as long as possible. It makes it harder when you can't put generic type parameters on, like in an inner function, but if `finalContinuation` returned a `'ret` then the types constrain you further and it's harder to get it wrong. I'm not sure whether that's possible here.